### PR TITLE
Rewrite logger system

### LINKED ISSUE #12082: embedded subcommand not working in Linux
- Gitea version (or commit ref): 1.12.1 (till latest)

## Description

The embedded subcommand does not work list or extract any files. 

## Screenshots

```
bash-5.0# gitea -v
Gitea version 1.12.1 built with GNU Make 4.2.1, go1.14.4 : bindata, sqlite, sqlite_unlock_notify
bash-5.0# gitea embedded list '**'
bash-5.0#
```

## Reason

Gitea `embedded` can not run with mismatched Linux user in an installed config, there will be no output at all.

## Workaround

Run gitea with the correct Linux user (as in the `app.ini`), using `su` or `sudo`

### PR DESCRIPTION
**Pull request by wxiaoguang**
## ⚠️ Breaking

The `log.<mode>.<logger>` style config has been dropped. If you used it, please check the new config manual & app.example.ini to make your instance output logs as expected.

Although many legacy options still work, it's encouraged to upgrade to the new options.

The SMTP logger is deleted because SMTP is not suitable to collect logs.

If you have manually configured Gitea log options, please confirm the logger system works as expected after upgrading.

## Description

Close #12082 and maybe more log-related issues, resolve some related FIXMEs  in old code (which seems unfixable before)

Just like rewriting queue #24505 : make code maintainable, clear legacy bugs, and add the ability to support more writers (eg: JSON, structured log)

There is a new document (with examples): `logging-config.en-us.md`

This PR is safer than the queue rewriting, because it's just for logging, it won't break other logic.

## The old problems

The logging system is quite old and difficult to maintain:
* Unclear concepts: Logger, NamedLogger, MultiChannelledLogger, SubLogger, EventLogger, WriterLogger etc
    * Some code is diffuclt to konw whether it is right: `log.DelNamedLogger("console")` vs `log.DelNamedLogger(log.DEFAULT)` vs `log.DelLogger("console")`
* The old system heavily depends on ini config system, it's difficult to create new logger for different purpose, and it's very fragile.
* The "color" trick is difficult to use and read, many colors are unnecessary, and in the future structured log could help
* It's difficult to add other log formats, eg: JSON format
* The log outputer doesn't have full control of its goroutine, it's difficult to make outputer have advanced behaviors
* The logs could be lost in some cases: eg: no Fatal error when using CLI.
* Config options are passed by JSON, which is quite fragile.
* INI package makes the KEY in `[log]` section visible in `[log.sub1]` and `[log.sub1.subA]`, this behavior is quite fragile and would cause more unclear problems, and there is no strong requirement to support `log.<mode>.<logger>` syntax.

## The new design

See `logger.go` for documents.

## Screenshot

<details>

image

image

image

</details>

## TODO

* [x] add some new tests
* [x] fix some tests
* [x] test some sub-commands (manually ....)

### CODE REVIEW

File: models/migrations/base/tests.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ import (
 	"xorm.io/xorm"
 )
 
+// FIXME: this file shouldn't be in a normal package, it should only be compiled for tests
```
Comment (lunny) [inline models/migrations/base/tests.go:27]: Maybe we need a build flag to address that since test package functions can not be shared by other packages.

File: modules/util/rotatingfilewriter/writer.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
```
Comment (lunny) [inline modules/util/rotatingfilewriter/writer.go:1]: Move to `modules/log/rotatingfilewriter`?

File: modules/testlogger/testlogger.go [hunk 3]
```diff
@@ -13,7 +13,6 @@ import (
 	"testing"
```
Comment (lunny) [inline modules/testlogger/testlogger.go:13]: Moving to `modules/log/testlogger`?

File: modules/util/rotatingfilewriter/writer.go (hunk 2 above)
Comment (wxiaoguang) [reply modules/util/rotatingfilewriter/writer.go:1]: No, it also works with #24257

File: modules/testlogger/testlogger.go (hunk 3 above)
Comment (wxiaoguang) [reply modules/testlogger/testlogger.go:13]: No, that's unrelated change. Not related to the "rewriting logger system" itself.

File: models/migrations/base/tests.go (hunk 1 above)
Comment (wxiaoguang) [reply models/migrations/base/tests.go:27]: Unrelated, and this PR is still in early stage, WIP.

**Comment by silverwind** [comment]
Not sure if in-scope here, but https://github.com/jwalton/go-supportscolor is a great tool to automatically detect color support for logging. It's a port of `supports-color` from JS land.

**Comment by wxiaoguang** [comment]
Added some new tests, the CI has passes many times.

This PR is ready for review.

File: custom/conf/app.example.ini [hunk 4]
```diff
@@ -550,34 +549,32 @@ ENABLE = true
...
-;; Set the log "modes" for the router log (if file is set the log file will default to router.log)
-ROUTER = console
+;; Buffer length of the channel, keep it as it is if you don't know what it is.
+;BUFFER_LEN = 10000
 ;;
-;; The router will log different things at different levels.
+;; Sub logger modes, a single comma means use default MODE above, empty means disable it
+;logger.access.MODE=
+;logger.router.MODE=,
+;logger.xorm.MODE=,
```
Comment (lunny) [inline custom/conf/app.example.ini:568]: Maybe we can rename it as `logger.sql.MODE`?

Comment (wxiaoguang) [reply custom/conf/app.example.ini:568]: The logger itself was hard-coded to "xorm" in code.

If you agree to rename all GetLogger("xorm") in code, we can change it.

Do you prefer to keep it as is, or rename all?

File: modules/setting/database.go [hunk 5]
```diff
@@ -92,7 +92,7 @@ func loadDBSetting(rootCfg ConfigProvider) {
 	Database.MaxOpenConns = sec.Key("MAX_OPEN_CONNS").MustInt(0)
 
 	Database.IterateBufferSize = sec.Key("ITERATE_BUFFER_SIZE").MustInt(50)
-	Database.LogSQL = sec.Key("LOG_SQL").MustBool(true)
+	Database.LogSQL = sec.Key("LOG_SQL").MustBool(false)
```
Comment (lunny) [inline modules/setting/database.go:95]: So this is a break change?

File: custom/conf/app.example.ini (hunk 4 above)
Comment (lunny) [reply custom/conf/app.example.ini:568]: Since this PR removed the key `XORM`, I think we can rename it to another one.

File: modules/setting/database.go (hunk 5 above)
Comment (wxiaoguang) [reply modules/setting/database.go:95]: I do not think it is "breaking".

The "install.go" already disabled it by default:

```
cfg.Section("database").Key("LOG_SQL").SetValue("false") // LOG_SQL is rarely helpful
```

File: custom/conf/app.example.ini (hunk 4 above)
Comment (wxiaoguang) [reply custom/conf/app.example.ini:568]: Another question , is "sql" a proper name for this logger?

This logger is only used by "xorm", and not only SQLs, but it also includes other XORM messages.

Personally I think the name "xorm" is not bad.

File: cmd/dump.go [hunk 6]
```diff
@@ -172,10 +172,7 @@ func runDump(ctx *cli.Context) error {
 	outType := ctx.String("type")
 	if fileName == "-" {
 		file = os.Stdout
-		err := log.DelLogger("console")
-		if err != nil {
-			fatal("Deleting default logger failed. Can not write to stdout: %v", err)
-		}
+		setupConsoleLogger(log.FATAL, log.CanColorStderr, os.Stderr)
```
Comment (delvh) [inline cmd/dump.go:175]: I don't quite understand.
Isn't that the opposite of what it did previously?

**Comment by wxiaoguang** [comment]
Actually, the useful colors are all kept (and improved), see the screenshots in the description  (expand the details)

File: modules/setting/mailer.go [hunk 7]
```diff
@@ -198,7 +199,7 @@ func loadMailerFrom(rootCfg ConfigProvider) {
 		ips := tryResolveAddr(MailService.SMTPAddr)
 		if MailService.Protocol == "smtp" {
 			for _, ip := range ips {
-				if !ip.IsLoopback() {
+				if !ip.IP.IsLoopback() {
```
Comment (lunny) [inline modules/setting/mailer.go]: The change looks unrelated?

Comment (wxiaoguang) [reply modules/setting/mailer.go]: Somewhat related (the `setting` package testing), without this, the tests blocks for long time if the network timeouts.

Comment (wxiaoguang) [reply modules/setting/mailer.go]: I will revert this

Comment (wxiaoguang) [reply modules/setting/mailer.go]: Done:

* f54bfd5b8de248ac413973336dc2f727425d9a52
* https://github.com/go-gitea/gitea/pull/24751

**Comment by zeripath** [comment]
The SMTP logger has been used to great effect by several people - it can be appropriate for infrequently emitted logs.

**Comment by wxiaoguang** [comment]
> The SMTP logger has been used to great effect by several people - it can be appropriate for infrequently emitted logs.

If any people needs it, I will add it back at rc-0 stage.

File: modules/util/rotatingfilewriter/writer.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+	return rfw, nil
+}
+
+func (rfw *RotatingFileWriter) Write(b []byte) (int, error) {
+	if rfw.options.Rotate && ((rfw.options.MaximumSize > 0 && rfw.currentSize >= rfw.options.MaximumSize) || (rfw.options.RotateDaily && time.Now().Day() != rfw.openDate)) {
+		if err := rfw.DoRotate(); err != nil {
+			// This should be
+			// return 0, err
+			// but the old behaviour does not return. This may lead to other errors.
+			fmt.Fprintf(os.Stderr, "RotatingFileWriter: %s\n", err)
```
Comment (KN4CK3R) [inline modules/util/rotatingfilewriter/writer.go]: As this PR is breaking, could you change this too? I did not change it in the audit PR because that one should not be breaking.

File: modules/log/event_writer.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+// SPDX-License-Identifier: MIT
+
+package log
+
+import (
+	"fmt"
+)
+
+type EventWriter interface {
+	EventWriterBase
```
Comment (lunny) [inline modules/log/event_writer.go:15]: There is no extra methods added, why we need the two interfaces?

File: modules/util/rotatingfilewriter/writer.go (hunk 8 above)
Comment (wxiaoguang) [reply modules/util/rotatingfilewriter/writer.go]: Sure, I can try to improve it. ~~Just a moment.~~ Done fe67dc4a36a5342eef34868c270d2f2991c391b9

The log writer will use fallback (stderr) to output logs if the file writer returns error, then no log would be lost.

---

Actually, the log rotator code has other problems.

The code was as old as 2014 ..... maybe at that time it works fine. But now if we use two rotating file writers to write to a single file (eg: gitea.log), they just conflict, because each writer has its internal bytes counter.

IMO this problem should be resolved by something like "shared/managed rotating file writer", while we can leave the problem to future.


----

Does fe67dc4a36a5342eef34868c270d2f2991c391b9 look good to you?

**Comment by lunny** [comment]
> The SMTP logger has been used to great effect by several people - it can be appropriate for infrequently emitted logs.

You mean a high level like error or fatal?

File: modules/log/event_writer.go (hunk 9 above)
Comment (wxiaoguang) [reply modules/log/event_writer.go:15]: This is a framework consideration.

Although they share the same methods at the moment, they are not the same thing.

In the future, EventWriter may has its own interface methods to force developers to implement something.

And I added more comments in code 1bb5e96e6abbdd3cc920add7ddc822e7777bdfa5

**Comment by silverwind** [comment]
> IMO this problem should be resolved by something like "shared/managed rotating file writer", while we can leave the problem to future.

A modern application should not do log routing or rotation. As per https://12factor.net/logs:

> A twelve-factor app never concerns itself with routing or storage of its output stream. It should not attempt to write to or manage logfiles. Instead, each running process writes its event stream, unbuffered, to stdout

I think this is how already gitea works in at least docker contexts, but I see no reason not to make it the single logging option. Yes, it means there won't be multiple output streams anymore, but those are confusing because they make correlating log entries harder.

**Comment by wxiaoguang** [comment]
At the moment, to avoid breaking as much as possible, it's better to keep these options.

And these options are for another reason: some people want to record different logs separately, eg: normal gitea.log and for audit purpose access.log

File: modules/util/rotatingfilewriter/writer.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	rfw.fd = fd
+
+	finfo, err := fd.Stat()
+	if err != nil {
+		return err
+	}
+	rfw.currentSize = finfo.Size()
+	rfw.openDate = time.Now().Day()
```
Comment (wolfogre) [inline modules/util/rotatingfilewriter/writer.go]: I know it's copied from the old code, but why not:
```suggestion
	rfw.openDate = finfo.ModTime().Day()
```

Think about this:
- The config is `MaximumSize=0` and `RotateDaily=true`.
- The Gitea instance is scheduled to shut down at 22:00 and then start at 9:00 the next day.
- So the log file will never be rotated, right?

Comment (wxiaoguang) [reply modules/util/rotatingfilewriter/writer.go]: There is no difference. There is `os.OpenFile` above, so `ModTime` is now.

Comment (wolfogre) [reply modules/util/rotatingfilewriter/writer.go]: I don't think so. It could just open an existing file.

Comment (wxiaoguang) [reply modules/util/rotatingfilewriter/writer.go]: Oh yes, you are right. Here it only opens the file, so mtime is as before.

**Review by wolfogre — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

File: custom/conf/app.example.ini [hunk 11]
```diff
@@ -550,34 +549,32 @@ ENABLE = true
...
-;;
-;; Switch off the router log
-;DISABLE_ROUTER_LOG=false
+;; Print Stacktrace with logs (rarely helpful, do not set) Either "Trace", "Debug", "Info", "Warn", "Error", default is "None"
+;STACKTRACE_LEVEL = None
 ;;
-;; Set the log "modes" for the router log (if file is set the log file will default to router.log)
-ROUTER = console
+;; Buffer length of the channel, keep it as it is if you don't know what it is.
+;BUFFER_LEN = 10000
```
Comment (delvh) [inline custom/conf/app.example.ini:563]: Should we rename it now to `BUFFER_LENGTH`?
And we should perhaps improve the documentation, a `keep it as it is if you don't know what it is` in the help sheet is not helpful at all.

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 12]
```diff
@@ -835,76 +835,62 @@ Default templates for project boards:
...
-  - You must be very careful to ensure that this template does not throw errors or panics as this template runs outside of the panic/recovery script.
+  - You must be very careful to ensure that this template does not throw errors or panics as this template runs outside the panic/recovery script.
 - `REQUEST_ID_HEADERS`: **\<empty\>**: You can configure multiple values that are splited by comma here. It will match in the order of configuration, and the first match will be finally printed in the access log.
   - e.g.
   - In the Request Header:        X-Request-ID: **test-id-123**
   - Configuration in app.ini:     REQUEST_ID_HEADERS = X-Request-ID
   - Print in log:                 127.0.0.1:58384 - - [14/Feb/2023:16:33:51 +0800]  "**test-id-123**" ...
 
-### Log subsections (`log.name`, `log.name.*`)
+### Log subsections (`log.writer-mode-name`)
```
Comment (delvh) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]:
```suggestion
### Log subsections (`log.<writer name>`)
```

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 13]
```diff
@@ -835,76 +835,62 @@ Default templates for project boards:
...
+### Console log mode (`log.console`, or `MODE=console`)
 
 - For the console logger `COLORIZE` will default to `true` if not on windows or the terminal is determined to be able to color.
 - `STDERR`: **false**: Use Stderr instead of Stdout.
 
-### File log mode (`log.file`, `log.file.*` or `MODE=file`)
+### File log mode (`log.file`, or `MODE=file`)
 
-- `FILE_NAME`: Set the file name for this logger. Defaults as described above. If relative will be relative to the `ROOT_PATH`
+- `FILE_NAME`: Set the file name for this logger. Defaults to "gitea.log" (exception: access log defaults to "access.log"). If relative will be relative to the `ROOT_PATH`
```
Comment (delvh) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `FILE_NAME`: Set the file name for this logger. Defaults to `gitea.log` (exception: access log defaults to `access.log`). If relative will be relative to the `ROOT_PATH`
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    identifier: "logging-config"
+---
+
+# Logging Configuration
+
+The logging configuration of Gitea mainly consists of 3 types of components:
+
+- The `[log]` section for general configuration
+- `[log.<mode-name>]` sections for the configuration of different log writers to output logs, aka: "writer mode", the mode name is also used as "writer name".
+- `[log]` section could contain sub-loggers like`logger.<logger-name>.<CONFIG-KEY>`
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
```suggestion
- The `[log]` section can also contain sub-logger configurations following the key schema `logger.<logger-name>.<CONFIG-KEY>`
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+
+{{< toc >}}
+
+## Collecting Logs for Help
+
+To collect logs for help and issue report, see [Support Options]({{< relref "doc/help/support.en-us.md" >}}).
+
+## The `[log]` section
+
+Configuration of logging facilities in Gitea happen in the `[log]` section and it's subsections.
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
```suggestion
Configuration of logging facilities in Gitea happen in the `[log]` section and its subsections.
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+NB: the whole message or string doesn't need to completely match.
+
+Please note this expression will be run in the writer's goroutine but not the logging event goroutine.
+
+#### `FLAGS`
+
+`FLAGS` represents the preceding logging context information that is
+printed before each message. It is a comma-separated string set. The order of values does not matter.
+
+It is default to `stdflags` (Equal to `date,time,medfile,shortfuncname,levelinitial`)
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
```suggestion
It defaults to `stdflags` (= `date,time,medfile,shortfuncname,levelinitial`)
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+- `slow` routers will be logged at WARN
+- `failed` routers will be logged at WARN
+
+### The "XORM" logger
+
+To make XORM outputs SQL logs, the `LOG_SQL` in `[database]` section should also be set to `true`.
+
+### The "Access" logger
+
+The Access logger is a new logger for version 1.9. It provides a NCSA
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
```suggestion
The Access logger is a new logger since Gitea 1.9. It provides a NCSA
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+
+If desired the format of the Access logger can be changed by changing
+the value of the `ACCESS_LOG_TEMPLATE`.
+
+Please note, the access logger will log at `INFO` level, setting the
+`LEVEL` of this logger to `WARN` or above will result in no access logs.
+
+#### The ACCESS_LOG_TEMPLATE
+
+This value represent a go template. It's default value is:
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
```suggestion
This value represents a go template. Its default value is
```

File: docs/content/doc/administration/logging-config.en-us.md [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+```
```
Comment (delvh) [inline docs/content/doc/administration/logging-config.en-us.md]:
````suggestion
```tmpl
````

File: modules/setting/config_provider.go [hunk 20]
```diff
@@ -33,6 +33,58 @@ type ConfigProvider interface {
 	Save() error
 }
 
+// ConfigSectionKey only searches the keys in the given section, but it is O(n).
+// ini package has a special behavior:  with "[sec] a=1" and an empty "[sec.sub]",
+// then in "[sec.sub]", Key()/HasKey() can always see "a=1" because it always tries parent sections.
```
Comment (delvh) [inline modules/setting/config_provider.go:38]: :exploding_head:

File: modules/log/color_console.go [hunk 21]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2014 The Gogs Authors. All rights reserved.
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
Comment (delvh) [inline modules/log/color_console.go]: Not anymore.

File: modules/lfs/pointer.go [hunk 22]
```diff
@@ -113,15 +111,11 @@ func (p Pointer) RelativePath() string {
 	return path.Join(p.Oid[0:2], p.Oid[2:4], p.Oid[4:])
 }
 
-// ColorFormat provides a basic color format for a Team
-func (p Pointer) ColorFormat(s fmt.State) {
+func (p Pointer) LogString() string {
 	if p.Oid == "" && p.Size == 0 {
-		log.ColorFprintf(s, "<empty>")
-		return
+		return "<Pointer empty>"
```
Comment (delvh) [inline modules/lfs/pointer.go]:
```suggestion
		return "<empty LFSPointer>"
```

File: modules/log/manager.go [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	dump := map[string]any{}
+	for name, logger := range m.loggers {
+		loggerDump := map[string]any{
+			"IsEnabled":    logger.IsEnabled(),
+			"EventWriters": logger.DumpWriters(),
+		}
+		dump[name] = loggerDump
```
Comment (delvh) [inline modules/log/manager.go:103]: Sounds like we should define an appropriate interface?

File: modules/log/event_writer_console.go [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+func NewEventWriterConsole(name string, mode WriterMode) EventWriter {
+	w := &eventWriterConsole{EventWriterBaseImpl: NewEventWriterBase(name, "console", mode)}
+	opt := mode.WriterOption.(WriterConsoleOption)
+	if opt.Stderr {
+		w.OutputWriteCloser = nopCloser{os.Stderr}
+	} else {
+		w.OutputWriteCloser = nopCloser{os.Stdout}
+	}
+	return w
+}
```
Comment (delvh) [inline modules/log/event_writer_console.go:36]:
```suggestion
type nopCloser struct {}

func (nopCloser) Close() error { return nil }

func NewEventWriterConsole(name string, mode WriterMode) EventWriter {
	return &eventWriterConsole{EventWriterBaseImpl: NewEventWriterBase(name, "console", mode), OutputWriteCloser: nopCloser{}}
}
```
Should be equivalent, right?
Or do you use the Writer for anything else?

File: models/perm/access/repo_permission.go [hunk 25]
```diff
@@ -102,45 +102,28 @@ func (p *Permission) CanWriteIssuesOrPulls(isPull bool) bool {
...
-			config := ""
-			if unit.Config != nil {
-				configBytes, err := unit.Config.ToDB()
-				config = string(configBytes)
-				if err != nil {
-					config = err.Error()
-				}
+func (p *Permission) LogString() string {
+	format := "<Permission AccessMode=%s, %d Units, %d UnitsMode(s): [ "
+	args := []any{p.AccessMode.String(), len(p.Units), len(p.UnitsMode)}
```
Comment (delvh) [inline models/perm/access/repo_permission.go:107]:
```suggestion
	args := make([]any, 0, 3 + 5*len(p.Units) + 2*len(p.UnitsMode))
	args = append(args, p.AccessMode.String(), len(p.Units), len(p.UnitsMode)}
```

File: cmd/dump.go (hunk 6 above)
Comment (wxiaoguang) [reply cmd/dump.go:175]: The new logic is right. 

If you have examples about bad cases, feel free to provide details.

Comment (delvh) [reply cmd/dump.go:175]: Yeah, my problem is simply that the console logger was previously deleted here, and now you set it up?
Isn't that the exact opposite?

File: custom/conf/app.example.ini (hunk 11 above)
Comment (wxiaoguang) [reply custom/conf/app.example.ini:563]: That's just the old document, this PR is large enough and get reviewed many times.

I also agree that there are still many legacy problems, they could be improved by future PRs.

File: docs/content/doc/administration/config-cheat-sheet.en-us.md (hunk 12 above)
Comment (wxiaoguang) [reply docs/content/doc/administration/config-cheat-sheet.en-us.md]: It also affects "mode", and the document was/is inconsistent, so we can keep the "writer mode name" here.

File: cmd/dump.go (hunk 6 above)
Comment (wxiaoguang) [reply cmd/dump.go:175]: No, the new logger only outputs to Stderr, it won't affect the dump output to Stdout.

File: docs/content/doc/administration/config-cheat-sheet.en-us.md (hunk 12 above)
Comment (delvh) [reply docs/content/doc/administration/config-cheat-sheet.en-us.md]: That's not my (primary) problem.
My main point is that it currently looks like there is a `log.writer-mode-name` section.
```suggestion
### Log subsections (`log.<writer-mode-name>`)
```

File: modules/log/color_console.go (hunk 21 above)
Comment (delvh) [reply modules/log/color_console.go]:
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
```

File: models/perm/access/repo_permission.go (hunk 25 above)
Comment (wxiaoguang) [reply models/perm/access/repo_permission.go:107]: According to many tests, the "pre-allocation" doesn't improve noticeable performance.

And here it doesn't affect performance, I prefer to keep readability instead of the unclear math trick.

File: modules/lfs/pointer.go (hunk 22 above)
Comment (delvh) [reply modules/lfs/pointer.go]: Alternatively
```suggestion
		return "<LFSPointer empty>"
```

File: modules/log/event_writer_console.go (hunk 24 above)
Comment (wxiaoguang) [reply modules/log/event_writer_console.go:36]: Sorry, I don't get the point "equivalent".

`nopCloser` means it writes, but it doesn't close.

File: modules/log/manager.go (hunk 23 above)
Comment (wxiaoguang) [reply modules/log/manager.go:103]: Sorry, I don't get the point about what kind of "interface".

This dump is for debugging and display purpose only.

File: modules/log/color_console.go (hunk 21 above)
Comment (wxiaoguang) [reply modules/log/color_console.go]: IIRC, according to LICENSE usage, the copyright head could be kept if the code comes from another file.

File: modules/log/event_writer_console.go (hunk 24 above)
Comment (delvh) [reply modules/log/event_writer_console.go:36]: I didn't find a `write` method, to me it looks as if its only method is `Close`.
And for that, we don't need to store the writer.
Where exactly is its `write` method?
In the `io.Writer` that doesn't seem to be used anywhere?
Or where?

File: modules/lfs/pointer.go [hunk 26]
```diff
@@ -113,15 +111,11 @@ func (p Pointer) RelativePath() string {
...
 	}
-	log.ColorFprintf(s, "%s:%d",
-		log.NewColoredIDValue(p.Oid),
-		p.Size)
+	return fmt.Sprintf("<Pointer %s:%d>", p.Oid, p.Size)
```
Comment (wxiaoguang) [inline modules/lfs/pointer.go]:
```suggestion
	return fmt.Sprintf("<LFSPointer %s:%d>", p.Oid, p.Size)
```

File: modules/log/event_writer_console.go (hunk 24 above)
Comment (wxiaoguang) [reply modules/log/event_writer_console.go:36]: That's Golang standard.

```
type nopCloser struct {
	io.Writer
}

w :=  nopCloser{otherWriter}

// now w.Writer is otherWriter

```

Then `w.Write` equals to `w.Writer.Write`, which calls `otherWriter`

File: modules/log/color_console.go (hunk 21 above)
Comment (delvh) [reply modules/log/color_console.go]: IIRC, we keep the year the same for existing files and use the current year for new files (-> moved or newly created).

File: modules/log/manager.go (hunk 23 above)
Comment (delvh) [reply modules/log/manager.go:103]: Well, we always fill this map with exactly the same information: `enabled` and `event writers`.
That sounds like we can convert this map into an extra type and take advantage of the type system to me.

File: modules/log/color_console.go (hunk 21 above)
Comment (wxiaoguang) [reply modules/log/color_console.go]: I can take this change (I am neutral).

But in most cases, a moved (renamed) file doesn't change its year.

File: modules/log/manager.go (hunk 23 above)
Comment (wxiaoguang) [reply modules/log/manager.go:103]: This is for debugging/testing/display purpose only. It won't (and shouldn't) be used by any other code.

Using JSON/map is quite fragile. But this map is never seriously used, so no need to use "type system" to restrict it.


For example, `IsEnabled` is already on every "logger", developers should never use the content in this dumped map.

**Review by delvh — APPROVED** [review]

**Comment by yardenshoham** [comment]
@KN4CK3R would you like to block this?

File: models/db/log.go [hunk 27]
```diff
@@ -14,67 +14,62 @@ import (
...
-func (l *XORMLogBridge) Log(skip int, level log.Level, format string, v ...interface{}) error {
-	return l.logger.Log(skip+1, level, format, v...)
+func (l *XORMLogBridge) Log(skip int, level log.Level, format string, v ...interface{}) {
+	l.logger.Log(skip+1, level, format, v...)
 }
 
 // Debug show debug log
 func (l *XORMLogBridge) Debug(v ...interface{}) {
-	_ = l.Log(stackLevel, log.DEBUG, fmt.Sprint(v...))
+	l.Log(stackLevel, log.DEBUG, "%s", fmt.Sprint(v...))
```
Comment (KN4CK3R) [inline models/db/log.go:37]: Is `"%s"` needed because the result of `fmt.Sprint` could produce a `%...` string?

File: modules/log/logger_impl.go [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+	// handle LogStringer values
+	for i, v := range msgArgs {
+		if cv, ok := v.(*ColoredValue); ok {
+			if s, ok := cv.v.(LogStringer); ok {
+				cv.v = s.LogString()
+			}
+		} else if s, ok := v.(LogStringer); ok {
+			msgArgs[i] = s.LogString()
+		}
+	}
```
Comment (KN4CK3R) [inline modules/log/logger_impl.go]: How does this interface intercept "debug logging"? I often use `log.Error("%#v", whatever)` to dump all members of `whatever`. Does that still work or does it print the value from `LogString`?

File: modules/graceful/releasereopen/releasereopen.go [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+	r.counter++
+	r.releaseReopeners[r.counter] = rr
+
+	return func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+
+		delete(r.releaseReopeners, r.counter)
+	}
```
Comment (KN4CK3R) [inline modules/graceful/releasereopen/releasereopen.go]: This does not look alright. If you register two instances and cancel the first one, this will delete the second instance because `r.counter` was incremented by the second call.

File: models/db/log.go (hunk 27 above)
Comment (wxiaoguang) [reply models/db/log.go:37]: Yes, it is needed. `fmt.Sprint(v...)` may contain other `%xx`.

I think the old code is wrong.

File: modules/log/logger_impl.go (hunk 28 above)
Comment (wxiaoguang) [reply modules/log/logger_impl.go]: ~~Yes, `LogString` works with any string-like `%`~~

File: modules/graceful/releasereopen/releasereopen.go (hunk 29 above)
Comment (wxiaoguang) [reply modules/graceful/releasereopen/releasereopen.go]: Oops, good catch. forgot to write a test for it ....

Now it is: f985c0f277bd2733a59d5e4bc2a5efc45413fe95

File: modules/log/logger_impl.go (hunk 28 above)
Comment (wxiaoguang) [reply modules/log/logger_impl.go]: Hmm, if you mean you still want to use `%#v` to dump all fields but ignore the `LogString`, at the moment it's not possible if the "whatever" provides `LogString` method, it's expected to use `LogString` to provide all details .....

----

update: improved

Comment (wxiaoguang) [reply modules/log/logger_impl.go]: With 60079bad3e8309b92a43fc5562ec8d7469140b2c , the `%#v` is also supported now.

**Comment by delvh** [comment]
Can this PR be added to the merge queue now?

**Comment by yardenshoham** [comment]
Let's give @KN4CK3R a day, if there's no response - merge

**Review by KN4CK3R — APPROVED** [review]
Should be good 👍 I think we will see in the next release if we missed something.

**Comment by silverwind** [comment]
Docs build failed: https://github.com/go-gitea/gitea/actions/runs/5040058905/jobs/9038577074

File: modules/graceful/releasereopen/releasereopen.go [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var errs []error
+	for _, rr := range r.releaseReopeners {
+		if err := rr.ReleaseReopen(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
```
Comment (brechtvl) [inline modules/graceful/releasereopen/releasereopen.go:48]: I'm getting this build error with go 1.19:
```
modules/graceful/releasereopen/releasereopen.go:48:16: undefined: errors.Join
```

It seems this was added in 1.20, while the Gitea `go.mod` lists 1.19 as the go version.
https://pkg.go.dev/errors#Join

Comment (silverwind) [reply modules/graceful/releasereopen/releasereopen.go:48]: We once had a build step on CI that built on minimum supported go version, but it appears that may have gotten lost with the migration to GitHub actions.

Comment (wxiaoguang) [reply modules/graceful/releasereopen/releasereopen.go:48]: Using Go 1.20 is in the plan

* https://github.com/go-gitea/gitea/pull/24257#discussion_r1177201691

Comment (silverwind) [reply modules/graceful/releasereopen/releasereopen.go:48]: Gitea supports the last two go versions and Go 1.21 is still around 3 months off. If we change this policy, we have to update our docs etc.

Comment (wxiaoguang) [reply modules/graceful/releasereopen/releasereopen.go:48]: Hmm, yes, the problem is that if we still use Golang 1.19, then after Golang 1.21 gets released, Golang 1.19 won't get any updates.

So, I think it's good to make Gitea's release can use active Golang version in its lifecycle (it's not a must if most people prefer to keep Gitea 1.20 on Golang 1.19)

Comment (wxiaoguang) [reply modules/graceful/releasereopen/releasereopen.go:48]: I also remembered that somewhere mentions that "Gitea supports the last two go versions", but I can not find it now .... where is it? 🤣  Has it already been removed?

**Comment by silverwind** [comment]
Possible new issue from this: https://github.com/go-gitea/gitea/issues/24882

**Comment by wxiaoguang** [comment]
Fix flakey test in logger test #24883
